nltk/tag/perceptron.py

```diff
@@ -0,0 +1,317 @@
+        self.tagdict = {}
+        self.classes = set()
+        if load:
+            AP_MODEL_LOC = str(find('taggers/averaged_perceptron_tagger/'+PICKLE))
+            self.load(AP_MODEL_LOC)
+
+    def tag(self, tokens):
+        '''
+        Tag tokenized sentences.
+        :params tokens: list of word
+        :type tokens: list(str)
+        '''
+        prev, prev2 = self.START
+        output = []
+        
+        context = self.START + [self._normalize(w) for w in tokens] + self.END
+        for i, word in enumerate(tokens):
+            tag = self.tagdict.get(word)
+            if not tag:
+                features = self._get_features(i, word, context, prev, prev2)
+                tag = self.model.predict(features)
+            output.append((word, tag))
+            prev2 = prev
+            prev = tag
+
+        return output
+
+    def train(self, sentences, save_loc=None, nr_iter=5):
+        '''Train a model from sentences, and save it at ``save_loc``. ``nr_iter``
+        controls the number of Perceptron training iterations.
+
+        :param sentences: A list of (words, tags) tuples.
+        :param save_loc: If not ``None``, saves a pickled model in this location.
+        :param nr_iter: Number of training iterations.
+        '''
+        self._make_tagdict(sentences)
+        self.model.classes = self.classes
+        for iter_ in range(nr_iter):
+            c = 0
+            n = 0
+            for sentence  in sentences:
+                words = [word for word,tag in sentence]
+                tags  = [tag for word,tag in sentence]
+                
+                prev, prev2 = self.START
+                context = self.START + [self._normalize(w) for w in words] \
+                                                                    + self.END
+                for i, word in enumerate(words):
+                    guess = self.tagdict.get(word)
+                    if not guess:
+                        feats = self._get_features(i, word, context, prev, prev2)
+                        guess = self.model.predict(feats)
+                        self.model.update(tags[i], guess, feats)
+                    prev2 = prev
+                    prev = guess
+                    c += guess == tags[i]
+                    n += 1
+            random.shuffle(sentences)
+            logging.info("Iter {0}: {1}/{2}={3}".format(iter_, c, n, _pc(c, n)))
+        self.model.average_weights()
+        # Pickle as a binary file
+        if save_loc is not None:
+            pickle.dump((self.model.weights, self.tagdict, self.classes),
+                         open(save_loc, 'wb'), -1)
+        return None
+
+    def load(self, loc):
+        '''
+        :param loc: Load a pickled model at location.
+        :type loc: str 
+        '''
+        try:
+            w_td_c = pickle.load(open(loc, 'rb'))
+        except IOError:
+            msg = ("Missing trontagger.pickle file.")
+            raise Exception(msg)
```

Comment:
+1 to all that.

On Mon, Sep 28, 2015 at 6:36 AM, Mikhail Korobov <notifications@github.com>
wrote:

> In nltk/tag/perceptron.py
> <https://github.com/nltk/nltk/pull/1143#discussion_r40509395>:
>
> > +        # Pickle as a binary file
> > +        if save_loc is not None:
> > +            pickle.dump((self.model.weights, self.tagdict, self.classes),
> > +                         open(save_loc, 'wb'), -1)
> > +        return None
> > +
> > +    def load(self, loc):
> > +        '''
> > +        :param loc: Load a pickled model at location.
> > +        :type loc: str
> > +        '''
> > +        try:
> > +            w_td_c = pickle.load(open(loc, 'rb'))
> > +        except IOError:
> > +            msg = ("Missing trontagger.pickle file.")
> > +            raise Exception(msg)
>
> To clarify: I'd prefer not to handle this IOError at all.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nltk/nltk/pull/1143/files#r40509395>.
>